File: dart/lib/src/packets/src/server/ab.dart
part of '../../packets.dart';

class AbPacket extends ServerPacket {
  final List<BleData> devices;

  /// [AbPacket] is the device list packet.
  ///
  /// This packet is part of the packet sent from the server to the device.
  AbPacket({
    /// [devices] is the list of devices that are being ACKed.
    /// This is identified in the packet as `BLE+DATA`
    required this.devices,
  }) : super();

  /// [fromPacket] creates a [AbPacket] from a string packet in the format of `Layrz Protocol v3`.
  static AbPacket fromPacket(String raw) {
    if (!raw.startsWith('<Ab>') || !raw.endsWith('</Ab>')) {
      throw ParseException('Invalid identification packet, should be <Ab>...</Ab>');
    }

    final parts = raw.substring(4, raw.length - 5).split(';');
    int? receivedCrc = int.tryParse(parts[parts.length - 1], radix: 16);
    int? calculatedCrc = calculateCrc("${parts.sublist(0, parts.length - 1).join(';')};".codeUnits);

    if (receivedCrc != calculatedCrc) {
      throw CrcException('Invalid CRC, received: $receivedCrc, calculated: $calculatedCrc');
    }

    return AbPacket(devices: BleData.fromPackets(parts.sublist(0, parts.length - 1).join(';')));
  }

  /// [toPacket] returns the packet in the format of `Layrz Link Protocol v2`.
  @override
  String toPacket() {
    String payload = devices.map((e) => e.toPacket()).join(';');
    payload += ';';
    String crc = calculateCrc(payload.codeUnits).toRadixString(16).padLeft(4, '0').toUpperCase();
    return '<Ab>$payload$crc</Ab>';
  }

  @override
  String toString() => toPacket();

  @override
  AbPacket copyWith({
    List<BleData>? devices,
  }) {
    return AbPacket(devices: devices ?? this.devices);
  }
}

File: go/packets/server/ab.go
package server

import (
	"errors"
	"fmt"
	"strconv"
	"strings"

	"github.com/goldenm-software/layrz-protocol/go/v3/definitions"
	"github.com/goldenm-software/layrz-protocol/go/v3/internal/wire"
)

// Based on the `Layrz Protocol v2` specification. AbPacket is the Blueooth Low Energy packet
// sent from the server to the device
type AbPacket struct {
	// Is the list of devices that the server wants to send from the device
	Devices *[]definitions.BleData
}

// FromPacket is a method that converts a raw packet to a AbPacket
// based on the `Layrz Protocol v2` specification
//
// Returns an error if the packet is invalid
func (p *AbPacket) FromPacket(raw *string) error {
	if !strings.HasPrefix(*raw, "<Ab>") || !strings.HasSuffix(*raw, "</Ab>") {
		return errors.New("invalid package, should be <Ab>...</Ab>")
	}

	*raw = strings.TrimPrefix(*raw, "<Ab>")
	*raw = strings.TrimSuffix(*raw, "</Ab>")

	rawCrc := (*raw)[len(*raw)-4:]
	*raw = (*raw)[:len(*raw)-4]

	receivedCrc, err := strconv.ParseUint(rawCrc, 16, 16)

	if err != nil {
		return errors.New("cannot convert CRC to integer")
	}

	calculatedCrc := wire.Calculate([]byte(*raw))

	if calculatedCrc != uint16(receivedCrc) {
		return fmt.Errorf("invalid CRC, received: %04X, calculated: %04X", receivedCrc, calculatedCrc)
	}

	*raw = strings.TrimSuffix(*raw, ";")

	parts := strings.Split(*raw, ";")

	devices := make([]definitions.BleData, 0)

	for _, part := range parts {
		subparts := strings.Split(part, ":")
		if len(subparts) != 2 {
			return errors.New("invalid device definition")
		}

		macAddress := ""

		for j := 0; j < len(subparts[0]); j += 2 {
			macAddress += subparts[0][j : j+2]
			if j != len(subparts[0])-2 {
				macAddress += ":"
			}

		}

		devices = append(devices, definitions.BleData{
			MacAddress: &macAddress,
			Model:      &subparts[1],
		})

		p.Devices = &devices
	}

	return nil
}

// ToPacket is a method that converts a AbPacket to a raw packet
// based on the `Layrz Protocol v2` specification
func (p *AbPacket) ToPacket() *string {
	devices := make([]string, 0)

	for _, device := range *p.Devices {
		macAddress := strings.ReplaceAll(*device.MacAddress, ":", "")
		devices = append(devices, macAddress+":"+(*device.Model))
	}

	content := strings.Join(devices, ";")
	content += ";"

	crc := wire.Calculate([]byte(content))
	content += fmt.Sprintf("%04X", crc)

	content = fmt.Sprintf("<Ab>%s</Ab>", content)
	return &content
}
